lib/ui/recommendation.js

```diff
@@ -44,37 +48,71 @@ Recommendation.prototype = {
     // Note: the XUL DOM isn't updated to insert the
     // 'universal-search-recommendation' element until the popup is opened
     // once. So, poll for the element until it's ready, then attach listeners
-    // to it.
-    this._pollForElement();
+    // to it, and notify the HighlightManager that the recommendation has
+    // been created.
+    this._pollForElement(() => {
+      this.attachListeners();
+      this.events.publish('recommendation-created');
+    });
 
     this.events.subscribe('recommendation', this.show);
     this.events.subscribe('enter-key', this.navigate);
     this.events.subscribe('urlbar-change', this.hide);
     this.events.subscribe('before-popup-hide', this.hide);
   },
-
-  destroy: function() {
+  attachListeners: function() {
+    this.el.addEventListener('click', this.navigate);
+    this.el.addEventListener('mouseenter', this.onMouseEnter);
+    this.el.addEventListener('mousemove', this.onMouseMove);
+    this.el.addEventListener('mouseleave', this.onMouseLeave);
+  },
+  detachListeners: function() {
     this.el.removeEventListener('click', this.navigate);
+    this.el.removeEventListener('mouseenter', this.onMouseEnter);
+    this.el.removeEventListener('mousemove', this.onMouseMove);
+    this.el.removeEventListener('mouseleave', this.onMouseLeave);
+  },
+  destroy: function() {
+    this.detachListeners();
     this.events.unsubscribe('recommendation', this.show);
     this.events.unsubscribe('enter-key', this.navigate);
     this.events.unsubscribe('urlbar-change', this.hide);
     this.events.unsubscribe('before-popup-hide', this.hide);
 
+    this.win.clearTimeout(this.mouseMoveTimeout);
+
     delete this.el;
     delete this.win;
   },
-
-  _pollForElement: function() {
+  _pollForElement: function(cb) {
```

Comment:
Using a callback, not a promise, because we don't need the rejection option, and this is the least amount of code that works.